### DISCUSSION TITLE
Working around .GOO's limited abilities...

### DISCUSSION THREAD

**Discussion by bryanseger** [sn4k3/UVtools#761] [General]
To date, I have not been able to fully utilize UVtools on the new .GOO format for the Mars 4 Max and 4 DLP. I think it's entirely on Elegoo's side, but I think I have come up with a workaround specifically meant for exposure tests. 

On either printer, the build head moves the layer thickness amount regardless of what Z value UVtools says. I can work with this, but I need a change on UVtools first.

Right now, the minimum layer height within the 'Exposure time finder' module is .010mm. I want to make this .001mm instead so that I can make my workaround actually work. The thought here is that I'll do 25 exposure strips within one test. The printer will physically move .001mm between each strip, but it'll advance a total of .025mm between layers per strip, so that my exposure test for .025mm layer thicknesses is actually correct, if that makes sense....

So, can the minimum value here be changed to .001mm? I am not familiar with forking the program over and making my own tweaks...perhaps it's easy, I just don't know where to start. 

![image](https://github.com/sn4k3/UVtools/assets/4185392/1daa561f-d31f-42e8-ad4e-78652facfc5e)

**Comment by sn4k3** [comment]
> To date, I have not been able to fully utilize UVtools on the new .GOO format for the Mars 4 Max and 4 DLP.

You mean for multiple-exposure test right? Right now most formats have problems with it, simply firmwares are not ready to deal with such strategy of printing. We can only have this fully working with peace of mind on a true gcode printer, this should not be as painful as it is but it seems they refuse to adapt to have zeros on fields and listen to community needs, as long we are locked I can't see it going much further, we need mSLA printers to be at FDM levels in firmware department (Free and real open source).

> So, can the minimum value here be changed to .001mm? 

There's a good reason UVtools minimum is 0.010mm, it's because most printers are rated to 0.010mm at Z resolution, and even such I think it's an optimistic value, don't forget these printers are consumable toys to be forgotten within a year or less, not the real precise machines.
Thats also why UVtools suggest being inside 0.030 as minimum, some users using native slicer with Anycubic machines reported printing quality issues, their problem was printing at 0.025mm, they decrease resolution and problem solved! Printing at the edge of a told spec is always something I don't recommend.
Another reason is because you will "never" benefit from such resolutions and different resin's also have a different min/max layer thickness.

The recommendation:
![image](https://github.com/sn4k3/UVtools/assets/113281/188b555b-1821-44ed-8298-67da6cf09146)

The test you run should ran in same layer thickness you will print your real model otherwise is not valid.
You can however run a script to duplicate all layers x10 and set them all to 0.001mm inside the script, which IMO is asking for troubles and not the expected outcome.

**Reply by sn4k3** [reply]
If you just want to change all thickness without duplication of layers go to File - terminal and send this command:

`foreach(var layer in SlicerFile) layer.PositionZ /= 10f;`

That will offset layers positions by 0.1 (I'm supposing you generating test with 0.01mm thickness which will turn them into 0.001mm). You might also want to skip the bottom base layers

**Reply by bryanseger** [reply]
Thanks for the reply. I understand that printing at .001mm layers is not ideal for most printers, but in this case on the Mars 4 Max (and 4 DLP), if I were to run 25x exposure test strips all in one print, then each strip is effectively being printed at .025mm, which is the desired layer thickness at I'm ultimately going for. 

If I run the exposure test at a legit .025mm thickness, at 25x exposure strips, then each strip is effectively being printed at .625mm (25 x .025mm) layer thickness, which is absurdly thick. This is because each new layer image the printer sees, it's advancing the Z by .025mm, instead of going to the same Z height as prescribed by UVtools (I think the printer is ignoring layer values and going with global values). 

I sliced my 'dummy' file in Lychee at .001mm layers and then brought that into UVtools, but when I go to create the exposure test, I simply cannot select .001mm layer thickness. Is there a File-terminal command that'll get me what I'm needing?


Also, slightly related but not to the Mars 4 printers...your UVtools program can/is being used on much more sophisticated printers, where single-micron layer thicknesses are certainly doable! Allowing .001mm micron layer thicknesses in the exposure test region does have merit for $$$$ printers!

**Reply by sn4k3** [reply]
I revised the exported goo format again, and properties and flag to follow the layer table are correctly set, and most probably ignored by printer, the best thing to do here is trying to contact them and ask to revise this in their firmware. I'm tired of broken stuff outside my scope and getting workarounds for something that should work in first place is just annoying and not the way it should be. This is not just a feature for ME but if "advanced mode" doesn't work at all is a serious limitation for what you payed to. The most impressive factor is that kind of problems are extremely easy and fast to fix as issue a firmware update, but all these mSLA firmware makers are terrible in accomplishing that or listen to community, for now we are stuck with their short mindset and capping their machines. Some have great machines (hardware) turn in junk just because of a piece of bad software, which is unacceptable! A good firmware can turn a machine into gold but they can't see it.

https://github.com/sn4k3/UVtools/blob/master/documentation/Goo-Format-Spec-V1.2.pdf
![image](https://github.com/sn4k3/UVtools/assets/113281/8547e140-a976-4163-905d-c7ec18a69c6e)

UVtools accepts 0.001mm at its core but the UI dialogs are still constrained to 0.01mm as minimum. in next update I can lower that limit.

After the test is generated and if you want to set the global layer height to 0.001 go to file terminal and send: `SlicerFile.SuppressRebuildPropertiesWork(() => SlicerFile.LayerHeight = 0.001f);` That will keep layers configured as you did on dialog but set global layer height alone.
However, note if printer use that value, exposure time need to be adjusted, resin and/or machine may not be able to print at such thickness. I really don't see a use nor good outcome for this workaround...

There's also another way to bypass the UI, which is run the Operation via code using the terminal or script. That way you are not constrained by the UI inputs and it will allow as low as 0.001mm

**Reply by bryanseger** [reply]
Tiago, thanks for the assistance on this. The terminal code did work exactly, and I confirmed the printer's Z movements (without resin) prior to actually starting a print. The head did move reasonably predictably in .001mm increments as expected, so for every new layer on exposure strip #1 (for example) out of a total of 25 exposure strips, it was .025mm higher than the previous layer. This worked perfectly!

I have reached out to Elegoo twice over the last few months to bring this to their attention, but to no avail. Perhaps there's a language barrier there, or perhaps I'm the 0.1% of the users for these printers pushing them further than they were ever intended to go (and thus, little incentive on Elegoo's side to correct issues), but they have not fixed anything yet.

If you end up updating UVtools to allow for .001mm layer heights, it will certainly get used. Perhaps have a note like Lychee does where it strongly discourages such thin layer heights but ultimately still allows for it.

![image](https://github.com/sn4k3/UVtools/assets/4185392/e6d52a69-557e-4ff9-851c-4720163f434c)

**Reply by sn4k3** [reply]
But isn't only the last test of the strip your target? if only the last model of strip is the valid this is still a waste of resin and time. I wonder if is not easier and faster just to print single tests. I not aware of your full workflow, maybe you are removing the objects that you don't need making black layers for them, with an easy hack you can bypass that layers by issue a black layer and spare resin, also you can attempt to exposure them for 0.01s but I guess it won't work either because per layer data is not respected (At least positionZ is not).

> or perhaps I'm the 0.1% of the users for these printers

It doesn't matter, you pay for it you should have the features they announce. Worse is that they refer to it as open source 🤦 The terms these days are incorrectly used and abused, just like today everything is AI... Do a program that output "Hello World" and you have an AI program 🤣
Anyway don't expect a fast fix but keep the pressure... I lost the contact for who gave me the goo spec otherwise I would try.

> If you end up updating UVtools to allow for .001mm layer heights

I can do that on next update and lower the limit on the UI inputs

**Reply by bryanseger** [reply]
I made the below image to explain what I'm doing here. I am running 25x unique exposure times (in this case, 8s to 32s in 1s increments). Since the Mars 4 Max ignores the Z layer height and instead moves the nominal layer thickness between each and every layer, I am running 25x exposure times so that the next new layer on each exposure strip is effectively .025mm higher than the previous. 

For instance, the exposure test starts below at .001mm Z height. The next strip is at .002mm, then .003mm, and so on, until it's completed the first layer for all 25 exposure strips ending at .025mm Z height. All these values are in red boxes below. Then the second layer begins for each of the strips, so the first strip is now at .026mm Z height, .025mm higher than the previous layer, thus truly giving me exposure tests at my desired .025mm layer thickness. This continues until all strips are complete.

It's an absurdly long print for sure (17 hours I think, but I ran it overnight), but this did give me a wide range of exposure times and still allows me to use the highly useful UVtools exposure test on the Mars 4 Max (and I think this technique would work on the Mars 4 DLP as well).

I fear that the new Saturn 3 and Mars 4 and 4 Ultra all suffer from this .GOO issue, but we do not have those printers to test with. Hopefully this explains what I was doing and why I was asking what I was asking. Thanks again!

![image](https://github.com/sn4k3/UVtools/assets/4185392/5eab8189-1658-48df-9197-f82b030b1b40)
